### PR TITLE
Use active wait and perform extra toggle if needed.

### PR DESCRIPTION
**Pull request by sdalu**
Used the same technique as the nrf51-pwm-library in Nordic github

### CODE REVIEW

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 1]
```diff
@@ -63,19 +70,32 @@ PWMDriver PWMD3;
 /* Driver local variables and types.                                         */
 /*===========================================================================*/
 
+static const uint8_t pwm_margin_by_prescaler[] = {
+    (PWM_GPIOTE_DECISION_TIME +   0) >> 0,
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: Please use 2 spaces for indentation.

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 2]
```diff
@@ -63,19 +70,32 @@ PWMDriver PWMD3;
...
+    (PWM_GPIOTE_DECISION_TIME +   1) >> 1,
+    (PWM_GPIOTE_DECISION_TIME +   3) >> 2,
+    (PWM_GPIOTE_DECISION_TIME +   7) >> 3,
+    (PWM_GPIOTE_DECISION_TIME +  15) >> 4,
+    (PWM_GPIOTE_DECISION_TIME +  31) >> 5,
+    (PWM_GPIOTE_DECISION_TIME +  63) >> 6,
+    (PWM_GPIOTE_DECISION_TIME + 127) >> 7,
+    (PWM_GPIOTE_DECISION_TIME + 255) >> 8,
+    (PWM_GPIOTE_DECISION_TIME + 511) >> 9
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: can leave a ',' at the line ending so future adds don't need to change this line.

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 3]
```diff
@@ -63,19 +70,32 @@ PWMDriver PWMD3;
...
-  // Deal with PWM channels     
-  uint8_t n;
-  for (n = 0 ; n < pwmp->channels ; n++) {
-    if (pwmp->timer->EVENTS_COMPARE[n]) {
-      pwmp->timer->EVENTS_COMPARE[n] = 0;
-
-      if (pwmp->config->channels[n].callback != NULL) {
-	pwmp->config->channels[n].callback(pwmp);
+  uint8_t channel;
+  // Deal with PWM channels
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: ChibiOS/RT style forbids `//` line comments.

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 4]
```diff
@@ -63,19 +70,32 @@ PWMDriver PWMD3;
...
-      if (pwmp->config->channels[n].callback != NULL) {
-	pwmp->config->channels[n].callback(pwmp);
+  uint8_t channel;
+  // Deal with PWM channels       
+  for (channel = 0 ; channel < pwmp->channels ; channel++) {
+    if (pwmp->timer->EVENTS_COMPARE[channel]) {
+      pwmp->timer->EVENTS_COMPARE[channel] = 0;
+
+      if (pwmp->config->channels[channel].callback != NULL) {
+	pwmp->config->channels[channel].callback(pwmp);
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: indentation.

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 5]
```diff
@@ -284,41 +311,81 @@ void pwm_lld_enable_channel(PWMDriver *pwmp,
...
-  NRF_PPI->CH[ppi_channel[0]].EEP =
-      (uint32_t)&pwmp->timer->EVENTS_COMPARE[channel];
-  NRF_PPI->CH[ppi_channel[0]].TEP =
-      (uint32_t)&NRF_GPIOTE->TASKS_OUT[gpiote_channel];
-  NRF_PPI->CH[ppi_channel[1]].EEP =
-      (uint32_t)&pwmp->timer->EVENTS_COMPARE[pwmp->channels];
-  NRF_PPI->CH[ppi_channel[1]].TEP =
-      (uint32_t)&NRF_GPIOTE->TASKS_OUT[gpiote_channel];
-  NRF_PPI->CHENSET = ((1 << ppi_channel[0]) | (1 << ppi_channel[1]));
+  // Deal with corner case: 0% and 100%
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: All comments should use `/* ... */` style

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 6]
```diff
@@ -284,41 +311,81 @@ void pwm_lld_enable_channel(PWMDriver *pwmp,
...
+
+    // Check GPIOTE state
+    const bool gpiote = (NRF_GPIOTE->CONFIG[gpiote_channel] &
+		   GPIOTE_CONFIG_MODE_Msk) != GPIOTE_CONFIG_MODE_Disabled;
+
+    // GPIOTE is currently running
+    if (gpiote) {
+      uint32_t current;
+      while (true) {
+	pwmp->timer->TASKS_CAPTURE[PWM_GPIOTE_PPI_CC] = 1;
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: indentation

File: os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c [hunk 7]
```diff
@@ -284,41 +311,81 @@ void pwm_lld_enable_channel(PWMDriver *pwmp,
...
+    } else {
+      // Create GPIO Task
+      NRF_GPIOTE->CONFIG[gpiote_channel] =
+	(GPIOTE_CONFIG_MODE_Task << GPIOTE_CONFIG_MODE_Pos) |
+	((gpio_pin << GPIOTE_CONFIG_PSEL_Pos    ) & GPIOTE_CONFIG_PSEL_Msk    )|
+	((polarity << GPIOTE_CONFIG_POLARITY_Pos) & GPIOTE_CONFIG_POLARITY_Msk)|
+	((outinit  << GPIOTE_CONFIG_OUTINIT_Pos ) & GPIOTE_CONFIG_OUTINIT_Msk );
+
+      pwmp->timer->TASKS_CAPTURE[PWM_GPIOTE_PPI_CC] = 1;
+      if (pwmp->timer->CC[PWM_GPIOTE_PPI_CC] > width)
```
Comment (utzig) [inline os/hal/ports/NRF5/NRF51822/hal_pwm_lld.c]: indentation

**Comment by utzig** [comment]
Please configure your text editor to use two-spaces for indentation and no tabs or any other strange styling.

**Comment by sdalu** [comment]
Styling fixed

**Pull request merged**